### PR TITLE
bnb: exit selection when best_waste is 0

### LINKED ISSUE #18257: Suggested BNB Search Optimization
I wanted to open an issue to understand why `bnb_search` as [implemented](https://github.com/bitcoin/bitcoin/pull/10637) @achow101 does not terminate once an optimal search is found.  This behavior is evident I believe when inspecting the [test](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/test/coinselector_tests.cpp#L178) which sets the target to `5` and uses the utxo pool `4,3,2,1`.  since combinations of `4` are evaluated first, the `bnb_search` finds an optimal solution of `4,1`.  However, instead of terminating when the `best_waste = 0` (no better solution can be found), the search continues and instead finds another solution `2,3` which also has a `best_waste = 0`.  Why not terminate once an optimal solution is found instead of continuing to search?

### PR DESCRIPTION
**Pull request by achow101**
If we find a solution which has no waste, just use that. This solution
is what we would consider to be optimal, and other solutions we find
would have to also have 0 waste, so they are equivalent to the first
one with 0 waste. Thus we can optimize by just choosing the first one
with 0 waste.

Closes #18257

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

**Comment by fjahr** [comment]
Concept ACK on this as it is an improvement!

However, looking at this I had another thought: given we have two possible solutions with 0 waste, would we not want to minimize fees by selecting the one with fewer inputs? So, instead of stopping early we could still run through all the tries and select the best waste option with the least inputs. Actually this could be an additional breaker between any equal waste options.

**Comment by achow101** [comment]
> However, looking at this I had another thought: given we have two possible solutions with 0 waste, would we not want to minimize fees by selecting the one with fewer inputs? So, instead of stopping early we could still run through all the tries and select the best waste option with the least inputs. Actually this could be an additional breaker between any equal waste options.

Fewer inputs does not necessarily minimize fees. It depends on the exact scripts being used. Regardless, I don't think an explicit check for fees is needed. So long as `fee > long_term_fee`, the `waste` metric will include an amount that is proportional to the fee each input requires, so in the end, the selection that has the least waste also pays the least in transaction fees.

**Comment by instagibbs** [comment]
> So long as fee > long_term_fee, the waste metric will include an amount that is proportional to the fee each input requires, so in the end, the selection that has the least waste also pays the least in transaction fees.

just to nit, in this case, `fee == long_term_fee`, right? I think it could go either way, if the user expects long term fees to go up or down they might want to pick more or fewer inputs. In the absence of evidence, simply returning with first result seems optimal.

**Comment by instagibbs** [comment]
aka concept ACK

File: src/wallet/test/coinselector_tests.cpp [hunk 1]
```diff
@@ -176,8 +176,8 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
     selection.clear();
 
     // Select 5 Cent
-    add_coin(3 * CENT, 3, actual_selection);
-    add_coin(2 * CENT, 2, actual_selection);
+    add_coin(4 * CENT, 4, actual_selection);
```
Comment (instagibbs) [inline src/wallet/test/coinselector_tests.cpp]: why these changes?

Comment (achow101) [reply src/wallet/test/coinselector_tests.cpp]: There is an optimal result earlier in the selection so that changes the outcome here.

**Review by yancyribbens — APPROVED** [review]
LGTM

**Comment by laanwj** [comment]
Obvious concept ACK.
Is this a case that happens in practice though? I mean, is it worth special-casing, or doesn't it really matter to do a few more iterations in an unlikely 'optimal' case?

**Comment by yancyribbens** [comment]
> Is this a case that happens in practice though?

That's also a question I have.  My intuition is it's very unlikely to have an exact match in practice.  Also, I think If the cost of change is sufficiently small, its unlikely to have any match including an exact match.

**Comment by achow101** [comment]
I think this condition is fairly hard to hit and probably doesn't make much of a difference. I think that most wallets will hit the iteration limit in BnB as we do search for more solutions after finding the first one, so exiting early probably isn't helping that much anyways.

**Comment by instagibbs** [comment]
I think this extra logic is really easy to understand and maintainable, therefore should be merged(provided proper review) regardless or not how theoretical people find it.

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/18262/commits/9b5950db8683f9b4be03f79ee0aae8a780b01a4b

**Review by meshcollider — COMMENTED** [review]
utACK 9b5950db8683f9b4be03f79ee0aae8a780b01a4b
